### ISSUE TITLE
Replacing plots that use sources that get affected by callback bug

### ISSUE THREAD

**Issue by Jss7268** [bokeh/bokeh#7110]
I am running this test with Bokeh v0.12.10

The bug that I am experiencing is when trying to replace a plot by creating a new plot. I also have a tool that performs a js_callback that changes the data on a ColumnDataSource. After using the box select tool and clicking the button a few times (sometimes just 1 time), I get issues.

The behaviour that I expect is that the the tool always draws a line from corner to corner of the box select region **(the js callback stops working as the 'source' is changed and no longer can run `source.change.emit()`)**

```
import numpy
from bokeh.layouts import row, column
from bokeh.models.sources import ColumnDataSource
from bokeh.models.widgets import Button
from bokeh.plotting import figure, curdoc
from bokeh.models.tools import BoxSelectTool
from bokeh.models.callbacks import CustomJS


def callback(src):
    js_src = """
            var geometry = cb_data['geometry'];
            var new_data = {};
            ['x0', 'x1', 'y0', 'y1'].forEach(function(key) {
                new_data[key] = [geometry[key]];
            });
            console.log(source)
            source.data = new_data;
            source.change.emit();
        """

    return CustomJS(args=dict(source=src), code=js_src)


def add_tool(plot, source):
    plot.add_tools(BoxSelectTool(callback=callback(source)))


class Foo:

    def __init__(self):
        data = {
            'y': numpy.array(range(0, 5000)),
            'x': numpy.array(range(0, 5000))
        }
        self.source = ColumnDataSource(data)
        self.line_source = ColumnDataSource(
            {'x0': [], 'x1': [], 'y0': [], 'y1': []})

        fig = figure(y_axis_type='log', tools=[])
        fig.line('x', 'y', source=self.source)

        fig.segment('x0', 'y0', 'x1', 'y1',
                    source=self.line_source,
                    color="#FF00FF")

        add_tool(fig, self.line_source)

        but = Button(label="Replace Plot")

        but.on_click(self.replace_plot)

        self.column = column(but, fig)

    def replace_plot(self):
        fig3 = figure(y_axis_type='linear', tools=[])
        #  not recreating the data source yields more strange behaviour
        self.source = ColumnDataSource(self.source.data)
        fig3.line('x', 'y', source=self.source)

        fig3.segment('x0', 'y0', 'x1', 'y1',
                     source=self.line_source,
                     color="#FF00FF")
        add_tool(fig3, self.line_source)
        self.column.children[1] = fig3


c = Foo().column
curdoc().add_root(c)

```
**Running `console.log(source)` in the callback code will sometimes produce an erroneous output:**
Correct: `e {document: t, destroyed: t, change: t, transformchange: t, attributes: {…}, …}`
Incorrect: `{id: "75f15720-24ea-43bd-95bc-fbe84b624dd8", type: "ColumnDataSource"}` (When this is output, the cannot read 'emit' error shown below gets produced)

Below the error from the console that this produces. The error seems very random as to when it will appear, and every time I stop and restart bokeh with the `bokeh serve` command it will behave differently.

- Uncaught TypeError: Cannot read property 'emit' of undefined
    at eval (eval at <anonymous> (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:25), <anonymous>:10:23)
    at r.CustomJS.r.execute (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:25)
    at e.r.BoxSelectToolView.e._emit_callback (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:34)
    at e.r.SelectToolView.e._select (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:34)
    at e.r.BoxSelectToolView.e._do_select (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:34)
    at e.r.BoxSelectToolView.e._pan_end (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:34)
    at e.<anonymous> (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:2)
    at t.emit (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:2)
    at t.r.UIEvents.t.trigger (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:2)
    at t.r.UIEvents.t._trigger (VM7389 bokeh.min.js?v=ef6e00d155662bd375c6bec8886eba7f:2)

**Comment by bryevdv** [comment]
It's possible the old tool and callback are hanging on longer than they should. That's worth investigating. However, I would also suggest restructuring your usage, as I would consider this the of updating to be sub-optimal. Rather:

Don't *replace* the plot at all. By far the best known an tested and understood usage of Bokeh is to create a plot, etc. then *update* that plot by updating data sources. This keeps complicated operations on the Bokeh document to a minimum, and is also typically more performant (there are specialized code paths for updating just `source.data` that will not get used if you just blow away entire plots and replace them)

cc @mattpap

**Comment by Jss7268** [comment]
After some more investigating into the callback, I found that with the newly created figure that is using the `line_source` for callback, if the `line_source` ColumnDataSource is reconstructed so that the callback is using a 'fresh' source, then there doesn't seem to be any more issues with `source` in the js callback being incorrect. For my purposes, it works for me to recreate that data source.

The fix was adding a line of code to recreate the line_source
```
def replace_plot(self):
        fig3 = figure(y_axis_type='linear', tools=[])
        self.source = ColumnDataSource(self.source.data)
        self.line_source = ColumnDataSource(self.line_source.data) #added this line
        fig3.line('x', 'y', source=self.source)

        fig3.segment('x0', 'y0', 'x1', 'y1',
                     source=self.line_source,
                     color="#FF00FF")
        add_tool(fig3, self.line_source)
        self.column.children[1] = fig3
```

I think that there being a workaround (replacing the data sources) makes this issue less important because at least I know now that it can be done (though if this is the desired bokeh functionality, I don't know)

To address your suggestion for not replacing the plot; I found that recreating a figure works for changing the y_scale from LinearScale to LogScale dynamically, whereas after the figure was displayed, it seemed its scale could not be changed. The method that I am using is a workaround for [this issue](https://github.com/bokeh/bokeh/issues/6575)
